src/rust/engine/src/nodes.rs

```diff
@@ -166,37 +168,62 @@ impl WrappedNode for Select {
           task: task.clone(),
           entry: Arc::new(self.entry.clone()),
         }),
-        &tasks::Rule::Intrinsic(Intrinsic { product, input })
-          if product == context.core.types.snapshot && input == context.core.types.path_globs =>
+	&Rule::Intrinsic(Intrinsic { product, input })
+	  if product == types.directory_digest && input == types.input_file_content =>
+          {
+            let context = context.clone();
+            let store = context.core.store();
+
+            self.select_product(&context, types.input_file_content, "intrinsic")
+              .and_then(move |files_content: Value| {
+                let filename = externs::project_str(&files_content, "path");
+                let bytes: bytes::Bytes = externs::project_bytes(&files_content, "content").into();
+
+                let digest = store.store_file_bytes(bytes, false)
+                  .map_err(|e| throw(&e));
+
+                digest.and_then(move |digest: hashing::Digest| {
+                  let mut directory = bazel_protos::remote_execution::Directory::new();
+                  directory.mut_files().push({
+                    let mut file_node = bazel_protos::remote_execution::FileNode::new();
+                    file_node.set_name(filename);
+                    file_node.set_digest((&digest).into());
+                    file_node
+                  });
```

Comment:
So I didn't realize this when discussing this with you earlier, but it seems like it might be possible to actually keep all of the logic which uses the `bazel_protos` crate within the `fs` crate, since storing bytes is a pretty generic thing to want to do, and other subcrates might want to use it, for example.

I think it should be possible to expose a `store_file_with_name()` (or something) method in the `Store`, which could replace the above call to `store.store_file_bytes()` and remove the rest of this `and_then()`?

Reply:
changed